### PR TITLE
64 livechat and chat managers updates update

### LINKED ISSUE #77: Improve livechat using manager

### PR DESCRIPTION
**Pull request by haodn98**
Added manager for livechats,
Updated chat manager

### CODE REVIEW

**Review by parsival2022 — APPROVED** [review]

File: livechats/managers.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    @classmethod
+    def get_users_conversations(cls, participant):
+        query = {
+            "$or": [
+                {"initiator_id": participant.user_id},
+                {"receiver_id": participant.user_id},
+            ]}
+        conversations = cls.db.find(query, {"messages": 0})
+        conversation_list = cls.to_list(conversations)
+        conversation_list = conversation_list.reverse()
```
Comment (mehalyna) [inline livechats/managers.py]: The `list.reverse()` method reverses the list in place and returns `None`. This line will set `conversation_list` to `None`, which is likely not the intended behavior.
You may use `conversation_list.reverse()` without assignment to reverse the list in place, or `conversation_list = conversation_list[::-1]` to create a reversed copy of the list

File: livechats/managers.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        query = {"_id": ObjectId(conversation_id)}
+        existing_conversation = cls.get_document(query)
+        if existing_conversation:
+            return cls.id_to_string(existing_conversation)
+        return None
+
+    @classmethod
+    def upgrade_messages(cls, conversation_id, messages):
+        if len(messages) > 50:
+            for i in range(math.ceil(messages / 50)):
```
Comment (mehalyna) [inline livechats/managers.py]: `messages` is a list, and you cannot divide a list by an integer, which will raise a `TypeError`.
Use `len(messages)` instead to get the number of messages: `range(math.ceil(len(messages) / 50))`

File: livechats/managers.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                if len(messages) > 50:
+                    cls.db.update_one({"_id": ObjectId(conversation_id)}, {"$push": {"messages": messages[0:50]}})
```
Comment (mehalyna) [inline livechats/managers.py]: When adding multiple messages, you should use the `$each` operator inside `$push` to append each message in the list to the messages array.
You can modify the update statement to use: `{"$push": {"messages": {"$each": messages[0:50]}}}`

File: livechats/managers.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                    messages = messages[50::]
```
Comment (mehalyna) [inline livechats/managers.py]: The slicing syntax is slightly off. Use `messages = messages[50:]` to get the remaining messages after the first 50

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: livechats/managers.py (hunk 1 above)
Comment (haodn98) [reply livechats/managers.py]: fixed

File: livechats/managers.py (hunk 2 above)
Comment (haodn98) [reply livechats/managers.py]: fixed

File: livechats/managers.py (hunk 3 above)
Comment (haodn98) [reply livechats/managers.py]: added, thanks

File: livechats/managers.py (hunk 4 above)
Comment (haodn98) [reply livechats/managers.py]: fixed

**Review by mehalyna — APPROVED** [review]

**Review by ydmyterko — APPROVED** [review]

**Pull request merged**